app/views/apipie_dsl/apipie_dsls/help.html.erb

```diff
@@ -1,58 +1,340 @@
+
+<p><%= (_("While rendering the template in preview mode, the input either loads the value or uses a placeholder in case, the input is not ready for the rendering. If the template is rendered for real and the input is not ready, an error is raised. The readyness criteria differs per input type, e.g. for Fact inputs linked to host's fact with name \"cpu_count\" is ready when @host facts contains value for such fact.")) %> </p>
+
+<p><%= _('As it was mentioned above, different template types support different template inputs. See the table below for available combinations.') %></p>
+
+<table class="table table-condensed">
+  <thead>
+  <tr>
+    <th><%= _('Template type') %></th>
+    <th><%= _('Available input types') %></th>
+  </tr>
+  </thead>
+  <tbody>
+  <tr>
+    <td>Provisioning template</td>
+    <td>Fact value, Variable, Puppet parameter</td>
+  </tr>
+  <tr>
+    <td>Partition table</td>
+    <td>Fact value, Variable, Puppet parameter</td>
+  </tr>
+  <tr>
+    <td>Report template</td>
+    <td>User input</td>
+  </tr>
+  <tr>
+    <td>Job template</td>
+    <td>User input, Fact value, Variable, Puppet parameter</td>
+  </tr>
+  </tbody>
+</table>
+
+<h2 id="snippets"><%= _('Snippets') %></h2>
+
+<p><%= _('Snippets are special kind of templates that can be easily reused in multiple other templates. Imagine there is a bootstrap procedure to be run on every provisioned host. At the same time, there are multiple kickstart templates. To avoid duplication a snippet can be used. Assuming there is a snippet created with a name Bootstrap, following example will inject it in a template') %></p>
+
+<%= apipie_dsl_example(apipie_erb_wrap("snippet('Bootstrap')")) %>
+
+<h2 id="safe_mode"><%= _('Safe mode') %></h2>
+
+<p><%= _('Safe Mode is how we refer to the safety mechanism we have to prevent doing harmful things. It whitelists Ruby methods, that can be used in the template. When a safe mode protected method is called accidentally, rendering fails with the error that says "Safemode doesn\'t allow to access ...".') %></p>
+
+<p><%= _('When writing a template, prefer to use macros and variables, that are already available with Safe Mode enabled. If you are looking for some information that you are not able to retrieve via allowed methods or macros listed in this documentation, you can consider disabling the Safe Mode. Doing that, you will be able to access all internal application variables and methods directly, when the template is rendered.') %></p>
+
+<p><%= _('Be aware, this is dangerous and it can cause performance and security issues. Disabling safe mode allows any Ruby code to be executed on the server. Any application user that can edit the template can modify the template, so that when it is rendered, it spawns a shell process and run arbitrary command in it. Such commands would be ran under the "foreman" user and would still be restricted by SELinux policy. Disabling Safe Mode is only recommended in trusted environment.') %></p>
```

Comment:
` <p><%= _('Be aware, this is dangerous and it can cause performance and security issues. Disabling safe mode allows any Ruby code to be executed on the server. Any application user that can edit the template can modify the template, so that when it is rendered, it spawns a shell process and run arbitrary command in it. Such commands would be ran under the "foreman" user and would still be restricted by SELinux policy. Disabling Safe Mode is only recommended in trusted environment.') %></p>`

To: 

` <p><%= _('Be aware, this is dangerous and it can cause performance and security issues. Disabling safe mode allows any Ruby code to be executed on the server. Any application user that can edit the template can modify the template, so that when it is rendered, it spawns a shell process and runs an arbitrary command in it. These commands are run as the "foreman" user and are still restricted by SELinux policy. Disabling Safe Mode is only recommended in a trusted environment.') %></p>`